pkg/aliases/script/run.go

```diff
@@ -28,40 +29,67 @@ func (script *Script) Run(args []string, opt docker.RunOption) error {
 
 	command := posix.Shell(dockerCmdString)
 	command.Env = os.Environ()
-	command.Stdin = os.Stdin
+
+	info, err := os.Stdin.Stat()
+	if err != nil {
+		return err
+	}
+	if (info.Mode()&os.ModeNamedPipe) != 0 || info.Size() > 0 {
+		command.Stdin = os.Stdin
+	} else {
+		command.Stdin = nil
+	}
+
 	command.Stdout = os.Stdout
 
 	stderr, err := command.StderrPipe()
 	if err != nil {
 		return nil
 	}
 
-	if err := command.Start(); err != nil {
-		return err
-	}
+	fn := func() error {
+		if err := command.Start(); err != nil {
+			return err
+		}
 
-	scanner := bufio.NewScanner(stderr)
-	for scanner.Scan() {
-		line := scanner.Text()
-		// invalid argument error converts go error
-		if strings.HasPrefix(line, "invalid argument") {
-			r := regexp.MustCompile(`^invalid argument "(.*?)" for "(.*?)" flag: (.*)$`)
-			matches := r.FindStringSubmatch(line)
-			if len(matches) > 0 {
-				return util.FlagError(matches[1], matches[2], matches[3])
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
+			line := scanner.Text()
+			// invalid argument error converts go error
+			if strings.HasPrefix(line, "invalid argument") {
+				r := regexp.MustCompile(`^invalid argument "(.*?)" for "(.*?)" flag: (.*)$`)
+				matches := r.FindStringSubmatch(line)
+				if len(matches) > 0 {
+					return util.FlagError(matches[1], matches[2], matches[3])
+				}
 			}
+			// remove docker prefix
+			if strings.HasPrefix(line, "docker: Error response from daemon: ") {
+				fmt.Fprintln(os.Stderr, "aliases: error response from daemon: "+strings.TrimPrefix(line, "docker: Error response from daemon: "))
+				continue
+			}
+			// suppress guidance to help
+			if line == "See 'docker run --help'." {
+				continue
+			}
+			fmt.Fprintln(os.Stderr, line)
 		}
-		// remove docker prefix
-		if strings.HasPrefix(line, "docker: Error response from daemon: ") {
-			fmt.Fprintln(os.Stderr, "aliases: error response from daemon: "+strings.TrimPrefix(line, "docker: Error response from daemon: "))
-			continue
-		}
-		// suppress guidance to help
-		if line == "See 'docker run --help'." {
-			continue
+
+		return command.Wait()
+	}
+
+	if (info.Mode()&os.ModeNamedPipe) != 0 && info.Size() == 0 {
+		// FIXME: timeout get from command line arguments
+		if err := util.Timeout(3*time.Second, fn); err != nil {
```

Comment:
It is not possible to solve all cases by just checking the input.
For example, in Goland Run, there is always a named pipe and the size of the input is zero.

Reply:
If there is a better solution, should change this.